src/api.rs

```diff
@@ -18,73 +19,78 @@ pub mod client {
   ///
-  /// The output of the function is a serializable object that can be
-  /// passed as input to `randomness_sampling()`.
-  pub fn format_measurement(
+  /// The output of the function can be passed as input to `construct_randomness_request()`.
+  pub fn prepare_measurement(
     measurement: &[Vec<u8>],
     epoch: u8,
-  ) -> Result<RandomnessSampling, NestedSTARError> {
+  ) -> Result<RandomnessRequestState, NestedSTARError> {
     let nm = NestedMeasurement::new(measurement)?;
-    Ok(RandomnessSampling::new(&nm, epoch))
+    let rsf = RandomnessSampling::new(&nm, epoch);
+    Ok(RandomnessRequestState::new(rsf))
   }
 
-  /// In `sample_randomness`, the client uses the output of
-  /// `format_measurement()` to retrieve randomness for each layer of
-  /// their nested measurement. The randomness is retrieved via
-  /// a randomness fetcher instance, in coordination with a randomness server.
+  /// In `construct_randomness_request`, the client uses the output of
+  /// `prepare_measurement` to construct a JSON request to be sent to
+  /// the randomness server, to retrieve randomness for each layer of
+  /// the nested measurement.
   ///
-  /// The output of the function can be passed as an input to
-  /// `construct_message()`.
-  pub fn sample_randomness(
-    fetcher: &impl randomness::Fetcher,
-    rsf: RandomnessSampling,
-    verification_key: &Option<ppoprf::ServerPublicKey>,
-  ) -> Result<MessageGeneration, NestedSTARError> {
-    // construct randomness request state
-    let state = randomness::RequestState::new(rsf.clone());
-
-    // fetch randomness
-    let results = fetcher.fetch(state.request())?;
-    let finalized = state.finalize_response(&results, verification_key)?;
-
-    // output in client aggregation message generation format
-    MessageGeneration::new(rsf, finalized)
+  /// Each element of the output of the function can base64 encoded, and
+  /// sent to the randomness server.
+  pub fn construct_randomness_request(
+    rrs: &RandomnessRequestState,
+  ) -> Vec<Vec<u8>> {
+    rrs
+      .blinded_points()
+      .iter()
+      .map(|p| p.as_bytes().to_vec())
+      .collect()
   }
 
-  /// In `construct_message` the client uses the output from
-  /// `randomness_sampling()`, containing their nested measurement and
-  /// generated randomness and generates a JSON-formatted aggregation
-  /// message.
+  /// In `construct_message` the client uses the base64 decoded output from
+  /// the JSON response of the randomness server (the generated randomness) and the
+  /// output of `prepare_measurement` (containing the nested measurement/blinded points)
+  /// and generates a bincode-formatted aggregation message.
   ///
   /// The client can optionally specify any amount of additional data
   /// to be included with their message in `aux`.
   pub fn construct_message(
-    mgf: MessageGeneration,
+    randomness_points: &[&[u8]],
+    randomness_proofs: Option<&[&[u8]]>,
```

Comment:
I personally think that passing in the response directly from the randomness server `eval` function directly may be better here. This means that the client doesn't have to worry about extracting the points and proofs beforehand, instead these can be handled internally within this function.

Reply:
As discussed over our DM, reducing the amount of nested/wrapper structures being passed over the Rust/CXX bridge makes this easier to call. Since the caller is responsible for JSON/base64 parsing, it will have to extract the point and proof data anyway.

Reply:
Also, to justify letting the caller handle JSON parsing/base64 encoding, here is another quote from our DM: 

"Chromium already has functions for parsing/serializing JSON and base64. The compiled library is quite large, so i'm trying to eliminate redundancies i.e. having two JSON/base64 serializers in the browser. Also, the style of our API in PPOPRF is similar, the caller (star-randsrv) is responsible for constructing/parsing the randomness requests/responses."

Reply:
cc @rillian: Can I get your opinion on this? This sounds reasonable to me.

Reply:
brave-core already includes the base64 and serde_json crates, so the code-size/duplication argument doesn't apply. We do trust the Chromium equivalents, although all things being equal, parsing in Rust is better than in C++.

Making the caller responsible for the wire format is a good argument. It is, after all, aware of what randomness server implementation is plans to talk to and how. The advantage of requiring a particular serialization in an API is if we want to encourage that format for all users. It's hard to know if we're doing the correct thing for ourselves without seeing the corresponding client code, but I have no objection if you want to set it up this way.

Please use named types here if you're not going to pass around already-serialized buffers. Types like `Option<&[&[u8]]>` carry no validation state, so we have to carefully audit every use. From a safety point of view it really isn't much different from json.

Reply:
Interesting. I didn't consider that serde_json/base64 crates were shared among the libraries already used in the browser, but considering we compile everything in one big crate (via `build/rust/Cargo.toml`), this makes sense. We don't have request building/serialization/parsing/etc in PPRF, so keeping this API for consistency is a good reason IMO.

Is your comment regarding named types meant to be a rule of thumb for future library changes, or a comment on the current code? We pass the points and proofs into `process_randomness_response` which gives us `Vec<ppoprf::Evaluation>`.

Reply:
I guess I mean both in general, and the current code. Why `randomness_points: &[&[u8]]` and not `&[ppoprf::Point]`?

I'm less bothered by `randomness_proofs`, but if we're arguing for client control of serialization, why do the proofs have to be bincoded for nested-sta-rs? I know you're treating them as opaque in the wire protocol, but they're also just a pair or `RistrettoScalar` (`[u8; 32]`). That seems inconsistent.

Reply:
The client needs some way to pass the decoded data to the lib. The other option is performing the bincode decode in the FFI crate.